### ISSUE TITLE
Add support for Italian CIE 3.0

### ISSUE THREAD

**Issue by rcasta74** [OpenSC/OpenSC#2486]
Output of `pcsc_scan`:

```
ATR: 3B 8F 80 01 80 31 80 65 B0 85 04 00 11 12 0F FF 82 90 00 8A
+ TS = 3B --> Direct Convention
+ T0 = 8F, Y(1): 1000, K: 15 (historical bytes)
  TD(1) = 80 --> Y(i+1) = 1000, Protocol T = 0 
-----
  TD(2) = 01 --> Y(i+1) = 0000, Protocol T = 1 
-----
+ Historical bytes: 80 31 80 65 B0 85 04 00 11 12 0F FF 82 90 00
  Category indicator byte: 80 (compact TLV data object)
    Tag: 3, len: 1 (card service data byte)
      Card service data byte: 80
        - Application selection: by full DF name
        - EF.DIR and EF.ATR access services: by GET RECORD(s) command
        - Card with MF
    Tag: 6, len: 5 (pre-issuing data)
      Data: B0 85 04 00 11
    Tag: 1, len: 2 (country code, ISO 3166-1)
      Country code: 0F FF
    Tag: 8, len: 2 (status indicator)
      SW: 9000
+ TCK = 8A (correct checksum)

Possibly identified card (using /home/roberto/.cache/smartcard_list.txt):
3B 8F 80 01 80 31 80 65 B0 85 04 00 11 12 0F FF 82 90 00 8A
        Dutch driver licence (eID)
        Italian electronic identity card (eID)
        https://en.wikipedia.org/wiki/Italian_electronic_identity_card

```
Specifications: [cie_3.0_-_specifiche_chip](https://www.cartaidentita.interno.gov.it/downloads/2021/03/cie_3.0_-_specifiche_chip.pdf) (only in italian)

I would like to help but I don't know anything about smart card development; I have no idea how to get started. I tryed to look at the code of existing drivers, but I cannot understand how they work.

Reading the specifications it seams that it should be compliant to IAS ECC v1.0.1, so maybe the iasecc driver can be used in some way with this card as well.

**Comment by 3v1n0** [comment]
I'm looking at it in these days, this is definitely possible given that's based on `IAS/ECC v1.0.1` even though there are some aspects that have not been implemented in OpenSC yet, first of all the privacy protocol (DH RSA).

I'm playing with some initial experiments (I've setup the SM so far), even though I'd like how to manage few bits that are a bit different from other cards:
 - After SM setup, basically no operation is possible without entering PIN first, so i'm not sure how to handle this at application level (yet)
 - SM is mandatory, and even if it is different, it has various elements shared with sm-cwa14890, so I'm wondering how to make it default without using `opensc.conf` and without refactoring the whole `iasecc` to use `card->sm_ctx.ops.get_sm_apdu` (as per https://github.com/OpenSC/OpenSC/issues/2452#issuecomment-985991053)

It's also worth to mention that there are some official open source implementation, even though the quality of the code there is not always inspirational:
 - https://github.com/italia/cie-middleware
 - https://github.com/italia/cie-middleware-linux
 - https://github.com/italia/cieid-android-sdk

/cc @resoli

**Comment by frankmorgner** [comment]
Did you look at `card-itacns.c`? I'm wondering what the relation is to your Italian card.

Anyway, some application of your card uses PACE and EAC to establish SM. If this is how your PIN is verified, you can use `perform_pace` to setup and transparently use the secure messaging channel with DH.

**Comment by resoli** [comment]
Even @opensignature is doing some preliminary work: https://github.com/opensignature/OpenSC/commit/2dce5b8ddcc87e66362058d02771e456086f73ad

**Comment by resoli** [comment]
There is (somewhat, project is in a bad condition, don't look at the Readme) working code here: https://github.com/italia/cie-middleware-linux

**Comment by 3v1n0** [comment]
@frankmorgner it's completely different protocol from the old CNS, so should be instead based or integrated into `iasecc`. That's is what I'm doing so far, even though this is true only once the SM is setup.

However, currently `iasecc` to perform SM relies on smm-local module, which is not loaded unless there's a config file set (all the `iasecc_sm_*` calls relies on it), while the CIE would need that to be present all the time, so I'm wondering if going the _module way_ would be correct in this case.

**Comment by frankmorgner** [comment]
as said above, when it comes to PIN verification with PACE, you can use `perform_pace` without the need for configuring an SM module.

if you insist on the module, maybe @viktorTarasov can help.

**Comment by 3v1n0** [comment]
@frankmorgner I'm not insisting at all in using a module, but I would like to reuse the code that is already provided by smm-local (or better sm-cwa14890).

Otherwise another option is to re-implement that off-module, as this part of the auth is common to both cards with and without privacy protection

**Comment by opensignature** [comment]
Hi all, I am now entering and I do not know if "privacy protection" issue has already been discussed.
I don't think in OpenSC "Diffie-Hellman mutual authentication" (as described in IAS ECC v1.0.1 [1] pag. 61) is already implemented 

[1] [https://dvv.fi/documents/16079645/17324992/IAS+ECC+v1_0_1UK.pdf](https://dvv.fi/documents/16079645/17324992/IAS+ECC+v1_0_1UK.pdf)

**Comment by 3v1n0** [comment]
Nope, I've an initial (just the 1st step of the graph below) implementation (just code thrown there, not really organized or clean) in [here](https://github.com/3v1n0/OpenSC/commits/ita-cie-v3).

![image](https://user-images.githubusercontent.com/345675/149987072-3117f3c3-08ee-4ed1-8f2b-d7f05b2e2826.png)

But it requires multiple steps yet, so I was wondering how to mix that with the implementation that `iasecc` already provides (given that some parts can be shared).

**Comment by 3v1n0** [comment]
Ah, BTW... As per PACE, It's not clear how to do it as I don't find any documentation how to retrieve the `EF.CardAccess` as using the known AID's we're just getting errors.

**Comment by frankmorgner** [comment]
EF.CardAccess is in MF, that's why you can use `perform_pace` and be happy. For reference, you can use BSI TR-03110 or ICAO Doc 9303.

I'm just curious... why is step 1 performed in the picture above, if this uses unauthenticated keys on both sides? If I see this correctly, only Step 2-5 actually use authenticated keys from ICC/IFD.

**Comment by 3v1n0** [comment]
Oh, not sure why it wasn't finding it earlier while indeed I get the value now...

```
P:739115; T:0x140466351237120 02:18:52.558 [npa-tool] sm-eac.c:1299:perform_pace: 
EF.CardAccess (22 bytes):
31 14 30 12 06 0A 04 00 7F 00 07 02 02 04 01 01 1.0.............
02 01 02 02 01 02   
```

HOWEVER... it seems that it fails when using `pin`

```
P:739681; T:0x139884099545088 02:22:13.273 [npa-tool] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:739681; T:0x139884099545088 02:22:13.273 [npa-tool] apdu.c:539:sc_transmit: returning with: 0 (Success)
P:739681; T:0x139884099545088 02:22:13.274 [npa-tool] card.c:523:sc_unlock: called
P:739681; T:0x139884099545088 02:22:13.274 [npa-tool] reader-pcsc.c:740:pcsc_unlock: called
P:739681; T:0x139884099545088 02:22:13.274 [npa-tool] iso7816.c:128:iso7816_check_sw: Referenced data not found
P:739681; T:0x139884099545088 02:22:13.274 [npa-tool] sm-eac.c:1327:perform_pace: Could not select protocol properties (MSE: Set AT failed).
P:739681; T:0x139884099545088 02:22:13.274 [npa-tool] sm-eac.c:1536:perform_pace: returning with: -1216 (Data object not found)
```

Otherwise, it **_seemed_** to progress properly when using `mrz` but there was a failure due to the fact that we were computing a wrong APDU when passing `General authenticate (Map Nonce) command data  (264 bytes)` (I guess the card needed chaininig):

```
P:703556; T:0x140012767848448 01:40:51.758 [npa-tool] apdu.c:550:sc_transmit_apdu: called
P:703556; T:0x140012767848448 01:40:51.759 [npa-tool] apdu.c:261:sc_check_apdu: failed length check for short APDU
P:703556; T:0x140012767848448 01:40:51.759 [npa-tool] apdu.c:262:sc_check_apdu: APDU check failed ( (apdu->le > 256 || (apdu->lc > 255 && (apdu->flags & SC_APDU_FLAGS_CHAINING) == 0))) ...
P:703556; T:0x140012767848448 01:40:51.759 [npa-tool] apdu.c:323:sc_check_apdu: Invalid Case 4 short APDU:
cse=04 cla=10 ins=86 p1=00 p2=00 lc=264 le=256
resp=0x7ffe61e422c0 resplen=256 data=0x55e94e35a070 datalen=264
```

I was trying to fix the APDU, but starting from that I couldn't proceed anymore as it seems that the card is now "locked" when using PACE auth (still works when using BAC), and it disconnects all the times just after `eac_gen_auth_1_encrypted_nonce` is sent.:

```
P:739893; T:0x139741119854592 02:24:39.523 [npa-tool] sm-eac.c:786:eac_gen_auth_1_encrypted_nonce: 
General authenticate (Encrypted Nonce) command data (2 bytes):
7C 00 |.

P:739893; T:0x139741119854592 02:24:39.524 [npa-tool] apdu.c:550:sc_transmit_apdu: called
P:739893; T:0x139741119854592 02:24:39.524 [npa-tool] card.c:473:sc_lock: called
P:739893; T:0x139741119854592 02:24:39.524 [npa-tool] reader-pcsc.c:688:pcsc_lock: called
P:739893; T:0x139741119854592 02:24:39.524 [npa-tool] card.c:513:sc_lock: returning with: 0 (Success)
P:739893; T:0x139741119854592 02:24:39.524 [npa-tool] apdu.c:515:sc_transmit: called
P:739893; T:0x139741119854592 02:24:39.525 [npa-tool] apdu.c:363:sc_single_transmit: called
P:739893; T:0x139741119854592 02:24:39.525 [npa-tool] apdu.c:367:sc_single_transmit: CLA:10, INS:86, P1:0, P2:0, data(2) 0x55ba94f1c360
P:739893; T:0x139741119854592 02:24:39.525 [npa-tool] reader-pcsc.c:325:pcsc_transmit: reader 'Virtual PCD 00 01'
P:739893; T:0x139741119854592 02:24:39.525 [npa-tool] reader-pcsc.c:326:pcsc_transmit: 
Outgoing APDU (8 bytes):
10 86 00 00 02 7C 00 00 .....|..

P:739893; T:0x139741119854592 02:24:39.525 [npa-tool] reader-pcsc.c:244:pcsc_internal_transmit: called
P:739893; T:0x139741119854592 02:24:40.145 [npa-tool] reader-pcsc.c:332:pcsc_transmit: unable to transmit
P:739893; T:0x139741119854592 02:24:40.145 [npa-tool] apdu.c:380:sc_single_transmit: unable to transmit APDU: -1213 (Unknown data received from card)
P:739893; T:0x139741119854592 02:24:40.145 [npa-tool] apdu.c:518:sc_transmit: transmit APDU failed: -1213 (Unknown data received from card)
P:739893; T:0x139741119854592 02:24:40.145 [npa-tool] card.c:523:sc_unlock: called
P:739893; T:0x139741119854592 02:24:40.145 [npa-tool] reader-pcsc.c:740:pcsc_unlock: called
P:739893; T:0x139741119854592 02:24:40.146 [npa-tool] reader-pcsc.c:749:pcsc_unlock: Virtual PCD 00 01:SCardEndTransaction failed: 0x80100069
P:739893; T:0x139741119854592 02:24:40.146 [npa-tool] apdu.c:640:sc_transmit_apdu: sc_unlock failed
P:739893; T:0x139741119854592 02:24:40.146 [npa-tool] sm-eac.c:1342:perform_pace: Could not get encrypted nonce from card (General Authenticate step 1 failed).
P:739893; T:0x139741119854592 02:24:40.146 [npa-tool] sm-eac.c:1536:perform_pace: returning with: -1213 (Unknown data received from card)

```

It may be a security check, but I don't think there's a way to undo it. [Others mentioned the same](https://forum.italia.it/t/connessione-nfc-persa-durante-autenticazione-pace/10961/2).

So, tips are welcome, given that `--unblock` fails also with `Error: Data object not found`.

In any case, could be that in this card the key IDs have different indexes (I've the feeling that it only supports pin and mrz)?

**Comment by frankmorgner** [comment]
Before diving deeper into this...
1. please use the official software to read and use the card. Do this while capturing a PC/SC log so that we have something to compare the OpenSC commands with.
2. please read the specification to check *when* PACE is needed. Typically, the ePassport application is read with CAN/MRZ. However, the ePassport application doesn't contain any keys that can be used for signing or decryption in PKCS#11, for example. There are other cards that use PACE also for verifying the PIN in the eSign application, but that doesn't need to be the case for IASECC.

**Comment by resoli** [comment]
> ICAO Doc 9303.

https://www.icao.int/publications/pages/publication.aspx?docnum=9303

**Comment by resoli** [comment]
https://www.icao.int/publications/Documents/9303_p11_cons_en.pdf in particular for `BAC` and `PACE`

**Comment by 3v1n0** [comment]
> please use the official software to read and use the card. Do this while capturing a PC/SC log so that we have something to compare the OpenSC commands with.

There's no much point of doing it, given that the original software is open source (see the previous links).

> please read the specification to check when PACE is needed. Typically, the ePassport application is read with CAN/MRZ. However, the ePassport application doesn't contain any keys that can be used for signing or decryption in PKCS#11, for example

Yeah, in this case it seems that the only way to use PACE is via those two, and indeed provides limited infos, so as you said we need to take another route and can't use `perform_pace` for normal pkcs11 usage.

**Comment by frankmorgner** [comment]
>There's no much point of doing it, given that the original software is open source (see the previous links).

If I wanted to know if/when PACE is used, I find it way easier to read read some dozens APDUs than to read thousands lines of codes. Anyway, if you feel confident doing the latter, I'm fine with that.

**Comment by frankmorgner** [comment]
going back to the original topic... If you think your card actually implements the IASECC specification, do you actually need to change anything with regard to SM? After all, this card already has an sm module available...

**Comment by 3v1n0** [comment]
> I find it way easier to read read some dozens APDUs than to read thousands lines of codes. Anyway, if you feel confident doing the latter, I'm fine with that.

Yeah, but I was using your virtual smartcard reader so far, and it doesn't seem to work properly when using the card with the available official driver.

In any case there's no PACE stuff at all in the pkcs11 module, it's only used by some apps to retrieve cardholder infos (like birth place / date, picture...).
From what I see, there's no certificate exposed either. In any case, it can't be unlocked by PIN.

> If you think your card actually implements the IASECC specification, do you actually need to change anything with regard to SM? After all, this card already has an sm module available...

Yeah, as said OpenSC implements IASECC only partially, it does not handle the Privacy support (done as shown in the image above), which I've started to do so far.

My main concern was whether to reuse the smm-local module to handle this (and in this case, how to make sure that's loaded without having people to force edit their `opensc.conf`?) or instead handle it inside the driver, I think I'll probably do a mixture of the two (to reuse as much as code possible) though.

**Comment by resoli** [comment]
> pkcs11 module, it's only used by some apps to retrieve cardholder infos (like birth place / date, picture...).
> From what I see, there's no certificate exposed either. In any case, it can't be unlocked by PIN.

Are you sure? many people are currently using using the offical pkcs11 module via contactless reader configured in firefox ...
https://lists.debian.org/debian-italian/2021/05/msg00028.html

**Comment by 3v1n0** [comment]
Not sure I've been clear: you can currently use the official pkcs11 module in this way, it's not using PACE protocol at all from what I saw so far.
Also, you can't access certificate data when using PACE from what I saw so far (I need to double check, but PACE Is blocked here now, so need to try with BAC).

**Comment by frankmorgner** [comment]
In cie_3.0_-_specifiche_chip.pdf I can read that some files (certificates, serial number) are protected by both SM and PIN. So the designer of the card profile must have anticipated that the user is required to enter the PIN before any application can actually read the card. So to some extent, this behavior is intended.

both PKCS#11 on the application side and PKCS#15 on the card allow you specifying a certificate that *is* protected with a PIN. So in theory, such kind of profile is perfectly valid and usable. In practice, however, the first thing that applications do is they verify the PIN in order to access the certificate. So from a user's perspective a high level security operation (signing a document) is mixed with some management stuff to get things working (i.e. read out a certificate). That's a bad UX...

you can do the following: 
1. enable file caching in opensc.conf
2. write some tool which lets the user enter the PIN and reads out the whole card and caches the data in the process
3. now, whenever the files from the card should be read, the file cache is consulted, not the card. this allows reading the protected file from the disk instead of the card.

a different approach would be to cache the PIN somewhere persistently... but that obviously has some security implications...

**Comment by 3v1n0** [comment]
> enable file caching in opensc.conf

I was thinking about it, but that's so far applied to all the cards, no? I think we should allow to set it per card.

> write some tool which lets the user enter the PIN and reads out the whole card and caches the data in the process

How that would be invoked though? It must stay off OpenSC I suppose, no?
Like a tool to initialize the card for the user session?

> a different approach would be to cache the PIN somewhere persistently... but that obviously has some security implications...

This is what the official driver does, but that's indeed something I'd avoid.

Would `use_pin_caching` help though? I assume no from a first look...

**Comment by rcasta74** [comment]
If it can help, with the official driver, before I can use the card I need to enable it using a java application, which calls a function in the driver itself: [AbilitaCIE.cpp](https://github.com/italia/cie-middleware-linux/blob/master/cie-pkcs11/CSP/AbilitaCIE.cpp)
This function caches some data in a file along with the first half of the PIN: when I need to use the card (to access to a site with firefox, or to sign a pdf with okular) I'm still requested to insert the second half of the PIN.

**Comment by frankmorgner** [comment]
just change the global default of file caching to "enabled", we've discussed that elsewhere already. `pkcs15-tool --verify-pin` should be enough to initialize the file cache.

use_pin_caching only works within one process.

**Comment by 3v1n0** [comment]
>  what a horroble security theater! this combines the insecurity of a PIN leak on-disk with the inconvenience of letting the user enter the PIN every time the card is used.

Yeah... Definitely! Even if they pretend to "encrypt it", it's just reducing the security of the original PIN

**Comment by 3v1n0** [comment]
I've been trying PACE with another card... I'm still trying to understand what's wrong on the APDU we're sending...

@frankmorgner: any idea?

<details><summary>log</summary>

```
P:434351; T:0x140607303197248 17:10:59.700 [npa-tool] reader-pcsc.c:325:pcsc_transmit: reader 'Virtual PCD 00 01'
P:434351; T:0x140607303197248 17:10:59.700 [npa-tool] reader-pcsc.c:326:pcsc_transmit: 
Outgoing APDU (5 bytes):
00 B0 9C 00 DF .....

P:434351; T:0x140607303197248 17:10:59.700 [npa-tool] reader-pcsc.c:244:pcsc_internal_transmit: called
P:434351; T:0x140607303197248 17:10:59.762 [npa-tool] reader-pcsc.c:335:pcsc_transmit: 
Incoming APDU (24 bytes):
31 14 30 12 06 0A 04 00 7F 00 07 02 02 04 01 01 1.0.............
02 01 02 02 01 02 90 00                         ........

P:434351; T:0x140607303197248 17:10:59.762 [npa-tool] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:10:59.762 [npa-tool] apdu.c:539:sc_transmit: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:10:59.762 [npa-tool] card.c:523:sc_unlock: called
P:434351; T:0x140607303197248 17:10:59.762 [npa-tool] reader-pcsc.c:740:pcsc_unlock: called
P:434351; T:0x140607303197248 17:10:59.762 [npa-tool] sm-eac.c:1304:perform_pace: 
EF.CardAccess (22 bytes):
31 14 30 12 06 0A 04 00 7F 00 07 02 02 04 01 01 1.0.............
02 01 02 02 01 02                               ......

P:434351; T:0x140607303197248 17:10:59.913 [npa-tool] sm-eac.c:1320:perform_pace: Pace protocol 1223
P:434351; T:0x140607303197248 17:10:59.913 [npa-tool] sm-eac.c:636:format_mse_cdata: 
MSE command data (15 bytes):
80 0A 04 00 7F 00 07 02 02 04 01 01 83 01 02 ...............

P:434351; T:0x140607303197248 17:10:59.914 [npa-tool] apdu.c:550:sc_transmit_apdu: called
P:434351; T:0x140607303197248 17:10:59.914 [npa-tool] card.c:473:sc_lock: called
P:434351; T:0x140607303197248 17:10:59.914 [npa-tool] reader-pcsc.c:688:pcsc_lock: called
P:434351; T:0x140607303197248 17:11:04.121 [npa-tool] card.c:513:sc_lock: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:11:04.121 [npa-tool] apdu.c:515:sc_transmit: called
P:434351; T:0x140607303197248 17:11:04.122 [npa-tool] apdu.c:363:sc_single_transmit: called
P:434351; T:0x140607303197248 17:11:04.122 [npa-tool] apdu.c:367:sc_single_transmit: CLA:0, INS:22, P1:C1, P2:A4, data(15) 0x55b0fb8fb200
P:434351; T:0x140607303197248 17:11:04.122 [npa-tool] reader-pcsc.c:325:pcsc_transmit: reader 'Virtual PCD 00 01'
P:434351; T:0x140607303197248 17:11:04.122 [npa-tool] reader-pcsc.c:326:pcsc_transmit: 
Outgoing APDU (20 bytes):
00 22 C1 A4 0F 80 0A 04 00 7F 00 07 02 02 04 01 ."..............
01 83 01 02                                     ....

P:434351; T:0x140607303197248 17:11:04.122 [npa-tool] reader-pcsc.c:244:pcsc_internal_transmit: called
P:434351; T:0x140607303197248 17:11:04.235 [npa-tool] reader-pcsc.c:335:pcsc_transmit: 
Incoming APDU (2 bytes):
90 00 ..

P:434351; T:0x140607303197248 17:11:04.235 [npa-tool] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:11:04.235 [npa-tool] apdu.c:539:sc_transmit: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] card.c:523:sc_unlock: called
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] reader-pcsc.c:740:pcsc_unlock: called
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] sm-eac.c:791:eac_gen_auth_1_encrypted_nonce: 
General authenticate (Encrypted Nonce) command data (2 bytes):
7C 00 |.

P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] apdu.c:550:sc_transmit_apdu: called
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] card.c:473:sc_lock: called
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] reader-pcsc.c:688:pcsc_lock: called
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] card.c:513:sc_lock: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] apdu.c:515:sc_transmit: called
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] apdu.c:363:sc_single_transmit: called
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] apdu.c:367:sc_single_transmit: CLA:10, INS:86, P1:0, P2:0, data(2) 0x55b0fb8fb2a0
P:434351; T:0x140607303197248 17:11:04.236 [npa-tool] reader-pcsc.c:325:pcsc_transmit: reader 'Virtual PCD 00 01'
P:434351; T:0x140607303197248 17:11:04.237 [npa-tool] reader-pcsc.c:326:pcsc_transmit: 
Outgoing APDU (8 bytes):
10 86 00 00 02 7C 00 00 .....|..

P:434351; T:0x140607303197248 17:11:04.237 [npa-tool] reader-pcsc.c:244:pcsc_internal_transmit: called
P:434351; T:0x140607303197248 17:11:04.321 [npa-tool] reader-pcsc.c:335:pcsc_transmit: 
Incoming APDU (14 bytes):
7C 0A 80 08 44 DF 84 1B CC 6E 2E 5A 90 00 |...D....n.Z..

P:434351; T:0x140607303197248 17:11:04.321 [npa-tool] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:11:04.321 [npa-tool] apdu.c:539:sc_transmit: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:11:04.321 [npa-tool] card.c:523:sc_unlock: called
P:434351; T:0x140607303197248 17:11:04.321 [npa-tool] reader-pcsc.c:740:pcsc_unlock: called
P:434351; T:0x140607303197248 17:11:04.321 [npa-tool] sm-eac.c:801:eac_gen_auth_1_encrypted_nonce: 
General authenticate (Encrypted Nonce) response data (12 bytes):
7C 0A 80 08 44 DF 84 1B CC 6E 2E 5A |...D....n.Z

P:434351; T:0x140607303197248 17:11:04.321 [npa-tool] sm-eac.c:1351:perform_pace: 
Encrypted nonce from MRTD (8 bytes):
44 DF 84 1B CC 6E 2E 5A D....n.Z

P:434351; T:0x140607303197248 17:11:04.324 [npa-tool] sm-eac.c:875:eac_gen_auth_2_map_nonce: Adpu le 256, size is 256
P:434351; T:0x140607303197248 17:11:04.324 [npa-tool] sm-eac.c:879:eac_gen_auth_2_map_nonce: 
General authenticate (Map Nonce) command data (264 bytes):
7C 82 01 04 81 82 01 00 81 E1 D8 8B 3D 4C B0 EA |...........=L..
A6 83 82 53 FD B5 E4 AF B2 B1 A2 26 D8 23 D8 39 ...S.......&.#.9
A0 C4 09 C6 D2 D9 DD 7F FF C2 6C 6E 5F 00 C0 E5 ..........ln_...
68 7A 75 4D 17 DB B1 54 F1 44 44 A7 B1 5F A6 71 hzuM...T.DD.._.q
4B 90 62 00 3B 13 FF ED 12 C0 23 93 08 21 48 82 K.b.;.....#..!H.
[.... MORE STUFF ....]
AB 7B 7A 1B 55 0A 22 15 79 46 ED 96 FD 9F 51 CB .{z.U.".yF....Q.
98 3F E0 96 B9 93 EB C8 37 52 18 A7 EF FD 27 F6 .?......7R....'.
93 19 60 94 6F 61 7A A9 51 15 4C 29 5F 57 53 9B ..`.oaz.Q.L)_WS.
6C 14 90 B3 0D DF 19 BA                         l.......

P:434351; T:0x140607303197248 17:11:04.324 [npa-tool] apdu.c:550:sc_transmit_apdu: called
P:434351; T:0x140607303197248 17:11:04.324 [npa-tool] apdu.c:261:sc_check_apdu: failed length check for short APDU
P:434351; T:0x140607303197248 17:11:04.324 [npa-tool] apdu.c:262:sc_check_apdu: Erroring...
P:434351; T:0x140607303197248 17:11:04.324 [npa-tool] apdu.c:323:sc_check_apdu: Invalid Case 4 short APDU:
cse=04 cla=10 ins=86 p1=00 p2=00 lc=264 le=256
resp=0x7fff42f4fa50 resplen=256 data=0x55b0fb8fb4c0 datalen=264
P:434351; T:0x140607303197248 17:11:04.324 [npa-tool] sm-eac.c:1383:perform_pace: Could not exchange mapping data with card (General Authenticate step 2 failed).
P:434351; T:0x140607303197248 17:11:04.325 [npa-tool] sm-eac.c:1541:perform_pace: returning with: -1300 (Invalid arguments)
P:434351; T:0x140607303197248 17:11:04.326 [npa-tool] reader-pcsc.c:537:pcsc_reconnect: Reconnecting to the card...
P:434351; T:0x140607303197248 17:11:04.326 [npa-tool] reader-pcsc.c:362:refresh_attributes: Virtual PCD 00 01 check
P:434351; T:0x140607303197248 17:11:04.327 [npa-tool] reader-pcsc.c:387:refresh_attributes: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:11:04.327 [npa-tool] card.c:1227:match_atr_table: ATR     : 3b:8f:80:01:80:31:80:65:b0:85:04:00:11:12:0f:ff:82:90:00:8a
P:434351; T:0x140607303197248 17:11:04.327 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:00:70:0A:90:00:8B
P:434351; T:0x140607303197248 17:11:04.327 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.327 [npa-tool] card.c:1227:match_atr_table: ATR     : 3b:8f:80:01:80:31:80:65:b0:85:04:00:11:12:0f:ff:82:90:00:8a
P:434351; T:0x140607303197248 17:11:04.327 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:7F:96:00:00:00:31:B9:64:40:70:14:10:73:94:01:80:82:90:00
P:434351; T:0x140607303197248 17:11:04.327 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:7F:96:00:00:00:31:B8:64:40:70:14:10:73:94:01:80:82:90:00
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:8F:80:01:80:31:80:65:B0:85:04:00:11:12:0F:FF:82:90:00:8A
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1227:match_atr_table: ATR     : 3b:8f:80:01:80:31:80:65:b0:85:04:00:11:12:0f:ff:82:90:00:8a
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:FE:18:00:00:81:31:FE:45:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:FA
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:8E:80:01:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:18
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:DE:18:FF:81:91:FE:1F:C3:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:1C
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:DE:96:FF:81:91:FE:1F:C3:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:92
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:80:80:01:01
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:84:80:01:47:6f:49:44:00
P:434351; T:0x140607303197248 17:11:04.328 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:85:80:01:47:6f:49:44:00:00
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:86:80:01:47:6f:49:44:00:00:00
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:87:80:01:47:6f:49:44:00:00:00:00
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:88:80:01:47:6f:49:44:00:00:00:00:00
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:89:80:01:47:6f:49:44:00:00:00:00:00:00
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:8a:80:01:47:6f:49:44:00:00:00:00:00:00:00
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:8b:80:01:47:6f:49:44:00:00:00:00:00:00:00:00
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:8c:80:01:47:6f:49:44:00:00:00:00:00:00:00:00:00
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:8d:80:01:47:6f:49:44:00:00:00:00:00:00:00:00:00:00
P:434351; T:0x140607303197248 17:11:04.329 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.330 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:8e:80:01:47:6f:49:44:00:00:00:00:00:00:00:00:00:00:00
P:434351; T:0x140607303197248 17:11:04.330 [npa-tool] card.c:1241:match_atr_table: ignored - wrong length
P:434351; T:0x140607303197248 17:11:04.330 [npa-tool] card.c:1238:match_atr_table: ATR try : 3B:8f:80:01:47:6f:49:44:00:00:00:00:00:00:00:00:00:00:00:00
P:434351; T:0x140607303197248 17:11:04.330 [npa-tool] card.c:1245:match_atr_table: ATR mask: FF:FF:FF:FF:FF:FF:FF:FF:00:00:00:00:00:00:00:00:00:00:00:00
P:434351; T:0x140607303197248 17:11:06.596 [npa-tool] reader-pcsc.c:560:pcsc_reconnect: Virtual PCD 00 01:SCardReconnect returned: 0x00000000
P:434351; T:0x140607303197248 17:11:06.596 [npa-tool] card.c:414:sc_disconnect_card: called
P:434351; T:0x140607303197248 17:11:06.596 [npa-tool] reader-pcsc.c:673:pcsc_disconnect: Virtual PCD 00 01:SCardDisconnect returned: 0x00000000
P:434351; T:0x140607303197248 17:11:06.596 [npa-tool] card.c:436:sc_disconnect_card: returning with: 0 (Success)
P:434351; T:0x140607303197248 17:11:06.596 [npa-tool] ctx.c:930:sc_release_context: called
P:434351; T:0x140607303197248 17:11:06.597 [npa-tool] reader-pcsc.c:977:pcsc_finish: called
Error: Invalid arguments

```
</details>

**Comment by frankmorgner** [comment]
please try the following change:
```diff
diff --git a/src/libopensc/card.c b/src/libopensc/card.c
index 7b1d6f18d..5f5febaa7 100644
--- a/src/libopensc/card.c
+++ b/src/libopensc/card.c
@@ -80,11 +80,11 @@ void sc_format_apdu_cse_lc_le(struct sc_apdu *apdu)
                if (apdu->resplen <= SC_MAX_EXT_APDU_RESP_SIZE)
                        apdu->le = apdu->resplen;
                if (apdu->resplen && !apdu->datalen)
-                       apdu->cse = SC_APDU_CASE_2;
+                       apdu->cse = SC_APDU_CASE_2_EXT;
                if (!apdu->resplen && apdu->datalen)
-                       apdu->cse = SC_APDU_CASE_3;
+                       apdu->cse = SC_APDU_CASE_3_EXT;
                if (apdu->resplen && apdu->datalen)
-                       apdu->cse = SC_APDU_CASE_4;
+                       apdu->cse = SC_APDU_CASE_4_EXT;
        } else {
                /* short length */
                if (apdu->datalen <= SC_MAX_APDU_DATA_SIZE)
diff --git a/src/sm/sm-eac.c b/src/sm/sm-eac.c
index 397375fae..a21abf87d 100644
--- a/src/sm/sm-eac.c
+++ b/src/sm/sm-eac.c
@@ -691,7 +691,7 @@ static int eac_gen_auth_2_map_nonce(sc_card_t *card,
        EAC_GEN_AUTH_PACE_R *r_data = NULL;
        unsigned char *d = NULL, *p;
        int r, l;
-       unsigned char resp[SC_MAX_APDU_RESP_SIZE];
+       unsigned char resp[SC_MAX_EXT_APDU_RESP_SIZE];
 
        c_data = EAC_GEN_AUTH_PACE_C_new();
        if (!c_data) {
@@ -777,7 +777,7 @@ static int eac_gen_auth_3_perform_key_agreement(sc_card_t *card,
        EAC_GEN_AUTH_PACE_R *r_data = NULL;
        unsigned char *d = NULL, *p;
        int r, l;
-       unsigned char resp[SC_MAX_APDU_RESP_SIZE];
+       unsigned char resp[SC_MAX_EXT_APDU_RESP_SIZE];
 
        c_data = EAC_GEN_AUTH_PACE_C_new();
        if (!c_data) {
```

**Comment by 3v1n0** [comment]
Ok I will try, thanks.

However IIRC, the CIE doesn't support EXT APDU, but it uses chaining instead...

**Comment by frankmorgner** [comment]
Then feel free to set `apdu.flags |= SC_APDU_FLAGS_CHAINING` instead

**Comment by 3v1n0** [comment]
Yeah, that was the test I did before the initial crashes, iirc it wasn't working at the time but will try again.

**Comment by resoli** [comment]
@lgbaldoni don't know, but from next week I will be a CIE holdet, so will be back on this ...

**Comment by frankmorgner** [comment]
any update?

**Comment by opensignature** [comment]
I'm afraid nobody is really interested in adding CIE support in opensc.
I, @resoli, @3v1n0, and others have tried but it's not easy, especially for who dedicate only part of their free time to it.
In my opinion, until some of https://github.com/italia/cie-middleware-linux come in here, it will be difficult to continue this integration

**Comment by resoli** [comment]
Recently I became a CIE holder, so I'm able to do some trials on it. I tried to build the official middleware (https://github.com/italia/cie-middleware-linux), but without success. A derived voluntary project, https://github.com/M0Rf30/cie-middleware-linux has much more clear build instructions and better structure. Maybe @M0Rf30 would like to join to our efforts? I am really interested, but I have no deep knowledge of OpenSC, even if I had [some past experience](https://j4sign.sourceforge.io) with Smartcards. May be @3v1n0 and @opensignature could participate.

**Comment by M0Rf30** [comment]
> Recently I became a CIE holder, so I'm able to do some trials on it. I tried to build the official middleware (https://github.com/italia/cie-middleware-linux), but without success. A derived voluntary project, https://github.com/M0Rf30/cie-middleware-linux has much more clear build instructions and better structure. Maybe @M0Rf30 would like to join to our efforts? I am really interested, but I have no deep knowledge of OpenSC, even if I had [some past experience](https://j4sign.sourceforge.io) with Smartcards. May be @3v1n0 and @opensignature could participate.

Honestly what I've done is not something that involves development skills and I don't know the implementation details. 
What I've done is to use a modern build system and write simple definitions to compile sources without having an IDE, a big cleanup and re-organization of codebase with de-duplication. Take me in consideration in case of testing (both for physical ownership of an NFC reader and CIE).

**Comment by Fantu** [comment]
hi, today trying to use CIE on raspberry thinclient (using freerdp) I saw that CIE was not working, after with a search I saw that is not implemented in opensc yet and cie-middleware-linux is needed but don't build on armhf.
Someone can tell me if there is a fork that build on arm? or also only a pkcs11 module (that I suppose is enough for CIE working for smarcard share with freerdp, or I'm wrong?)
thanks for any reply

**Comment by rnhmjoj** [comment]
> cie-middleware-linux is needed but don't build on armhf.

It's [packaged](https://search.nixos.org/packages?channel=23.05&show=cie-middleware-linux&from=0&size=50&sort=relevance&type=packages&query=cie-middleware-linux) in NixOS, I confirmed it works on x86_64 and aarch64,

**Comment by M0Rf30** [comment]
> > cie-middleware-linux is needed but don't build on armhf.
> 
> It's [packaged](https://search.nixos.org/packages?channel=23.05&show=cie-middleware-linux&from=0&size=50&sort=relevance&type=packages&query=cie-middleware-linux) in NixOS, I confirm it works on x86_64 and aarch64, not sure about 32bit ARM.

from my side no intentions to support already deprecated architectures like 32bit ARM (armv6,armv5)

**Comment by frankmorgner** [comment]
Closing this issue due to inactivity. Please re-open the ticket if more input is available.

**Comment by lgbaldoni** [comment]
@M0Rf30 refactored the official middleware and created a stand-alone [pkcs11 library](https://github.com/M0Rf30/opencie-pkcs11).
Do you think this issue might be reconsidered?

Would it be challenging to turn said library into an opensc driver?

Also paging @ale-rinaldi.
